### ISSUE TITLE
Question about multi-token gold answer target alignment in RecursiveMAS

### ISSUE THREAD

**Issue by yuprincess** [RecursiveMAS/RecursiveMAS#22]
This seems a beautiful and interesting project.Thank you for releasing the paper and code. I am reading RecursiveMAS and have a question about the inner-loop training target.
On the project page / paper, the inner-loop objective is described as:
```text
L_in = 1 - cosine( R_in(H), Emb_theta_i(y) )
```
where Emb_theta_i(y) is described as the agent's own input-embedding distribution of the ground-truth answer y.

From the released code, I can see the adapter structure in modeling.py. The Adapter module applies LayerNorm, Linear, GELU, Linear, residual connection, and LayerNorm, while preserving the input sequence length:
```text
class Adapter(nn.Module):
    ...
    def forward(self, x):
        h = self.pre_ln(x)
        out = self.proj2(self.act(self.proj1(h)))
        out = x + out
        return self.post_ln(out)
```
My understanding is:
```text
latent states / memory slots:
R_in(H): [B, K, d]

gold answer embedding target:
Emb_theta_i(y): [B, T_y, d]
```
Here, K is the number of latent rollout steps, and T_y is the number of tokens in the gold answer.
My question is: when K != T_y, how are these two tensors aligned for the cosine loss in the actual training implementation?
For example, if the gold answer is:
```text
540 meters
```
and it is tokenized into two tokens, then Emb(y) would have length 2, while the latent memory may have a fixed length such as 4 or 8 slots.
In this case, do you use one of the following approaches?
```text
1. mean-pool Emb(y) over answer tokens, then expand it to K slots
2. mean-pool both R_in(H) and Emb(y) into single vectors
3. pad/truncate Emb(y) to match K
4. dynamically set K = T_y
5. another alignment method

```
I also want to confirm one more point:
```text
Is the target the static input embedding Emb_theta_i(y),
or is it a contextual hidden state such as:
Model(question + answer).last_hidden_state at the answer-token positions?
```
I would really appreciate your clarification. This would help me better understand and reproduce the training objective faithfully.
Thank you!

**Comment by jiaruzouu** [comment]
Hi [yuprincess](https://github.com/yuprincess),

Thanks for your kind words to our work! Regarding your questions:

In our training, the alignment is defined position-wise along the selected output-token trajectory. That is, for each selected generation position, the adapter output is compared with the corresponding target representation in the model’s embedding space. Among your listed options, it is closest to `option 5: another alignment method`, specifically position-level alignment over the selected generated-token positions.

For the target, we use the model’s own input embedding-space representation of the target tokens rather than a separately pooled answer vector. Conceptually, the goal is to make the inner adapter map the agent’s hidden states back into a representation compatible with the agent’s input embedding space, so that the latent states can be reused as meaningful internal “thoughts” for subsequent computation.

Hope this clarifies your questions!
